BrainPortal/app/controllers/tasks_controller.rb

```diff
@@ -771,6 +771,395 @@ def apply_operation(operation, taskids, do_in_spawn = false, options = {})
+  # the sandbox or published to the official zenodo site. Once
+  # published on the official zenodo site, it can't be reset.
+  def reset_zenodo
+    task_id = params[:id]
+    @task   = current_user.available_tasks.find(task_id)
+
+    if @task.zenodo_doi.present? && ! @task.zenodo_doi.starts_with?(ZenodoHelper::ZenodoSandboxDOIPrefix)
+      cb_error "This task's outputs have already been published and we cannot reset its publication state any more."
+    end
+
+    # Remember original IDs
+    orig_deposit_id         = @task.zenodo_deposit_id
+    orig_doi                = @task.zenodo_doi
+
+    # Reset task
+    @task.zenodo_deposit_id = nil
+    @task.zenodo_doi        = nil
+    @task.save_with_logging(current_user, [ :zenodo_deposit_id, :zenodo_doi ])
+
+    # Reset all files
+    zenodo_userfile_ids = @task.zenodo_outputfile_ids
+    @zenodo_userfiles   = zenodo_userfiles_from_ids(zenodo_userfile_ids)
+    @zenodo_userfiles.compact.each do |userfile|
+      next if userfile.zenodo_doi.present? && ! userfile.zenodo_doi.starts_with?(ZenodoHelper::ZenodoSandboxDOIPrefix)
+      userfile.zenodo_deposit_id = nil if userfile.zenodo_deposit_id == orig_deposit_id
+      userfile.zenodo_doi        = nil if userfile.zenodo_doi        == orig_doi
+      userfile.save_with_logging(current_user, [ :zenodo_deposit_id, :zenodo_doi ])
+    end
+
+    # Attempt to remove the deposit on Zenodo; if it's published
+    # it will not be deleted no matter how much we try.
+    zsite, deposit_id = (orig_deposit_id || "").split("-") # "main", "1234"
+    init_zenodo_client(zsite)
+    depo_api = ZenodoClient::DepositsApi.new
+    depo_api.delete_deposit(deposit_id.to_i) rescue 'ignore'
+
+    redirect_to :action => :zenodo
+  end
+
+  private
+
+  # Returns a list of userfiles out of the
+  # +ids+ ; nil elements are left in the array if
+  # any id matches a missing file or a file not
+  # accessible by the current user.
+  def zenodo_userfiles_from_ids(ids) #:nodoc:
+    ids.map do |id|
+      Userfile
+        .find_all_accessible_by_user(current_user)
+        .where(:id => id)
+        .first # we allow nils: it means the task returned an ID of a missing/unaccessible file
+    end
+  end
+
+  def zenodo_server_by_zsite(zsite) #:nodoc:
+    zsite == 'main' ?
+      'zenodo.org'  :
+      'sandbox.zenodo.org'
+  end
+
+  def token_by_zsite(zsite) #:nodoc:
+    token = zsite == 'main'             ?
+      current_user.zenodo_main_token    :
+      current_user.zenodo_sandbox_token
+    token.presence.try(:strip)
+  end
+
+  def init_zenodo_client(zsite) #:nodoc:
+    zsite  = zsite == 'main' ? 'main' : 'sandbox' # sanitize
+    token  = token_by_zsite(zsite)
+    server = zenodo_server_by_zsite(zsite)
+    ZenodoClient.configure do |config|
+      config.api_key['access_token'] = token
+      config.host                    = server
+      #config.debugging               = true # will print info on stdout; WARNING binary junk!
```

Comment:
Could be removed?

Reply:
No, this is really useful, and someone who doesn't know how the ZenodoClient library works would probably be happy to know how to turn on debugging. I was using it all the time while developing, and I can imagine wanting to uncomment it when doing future work.